src/function/sequence_functions.cpp

```diff
@@ -0,0 +1,125 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// sequence_functions.cpp
+//
+// Identification: src/function/sequence_functions.cpp
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "function/sequence_functions.h"
+
+#include "common/macros.h"
+#include "executor/executor_context.h"
+#include "catalog/catalog.h"
+#include "catalog/database_catalog.h"
+#include "catalog/sequence_catalog.h"
+#include "concurrency/transaction_context.h"
+#include "concurrency/transaction_manager_factory.h"
+#include "type/value_factory.h"
+
+namespace peloton {
+namespace function {
+
+/*@brief   The actual implementation to get the incremented value for the specified sequence
+ * @param   sequence name, executor context
+ * @return  the next value for the sequence
+ * @exception the sequence does not exist
+ */
+uint32_t SequenceFunctions::Nextval(executor::ExecutorContext &ctx,
+                                   const char *sequence_name) {
+  PELOTON_ASSERT(sequence_name != nullptr);
+  concurrency::TransactionContext* txn = ctx.GetTransaction();
+  // get the database oid for this transaction
+  oid_t database_oid = catalog::Catalog::GetInstance()
+          ->GetDatabaseObject(ctx.GetDatabaseName(), txn)->GetDatabaseOid();
+
+  // initialize a new transaction for incrementing sequence value
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  auto mini_txn = txn_manager.BeginTransaction();
+
+  // evict the old cached copy of sequence
+  txn->catalog_cache.EvictSequenceObject(sequence_name,database_oid);
+  auto sequence_object =
+          catalog::Catalog::GetInstance()
+                  ->GetSystemCatalogs(database_oid)
+                  ->GetSequenceCatalog()
+                  ->GetSequence(database_oid, sequence_name, mini_txn);
+  if (sequence_object != nullptr) {
```

Comment:
The function EvictSequenceObject() grantees to evict the given database_oid. So I think you don't need to search for sequence_object and double check again. Or, since EvictSequenceObject() will return false and true, you can use this information for help you make decision.

Reply:
The sequence may exist in the catalog but not exist in the cache. In this case, EvictSequenceObject() will return false but we want to call nextval on the sequence.